src/stdlib_experimental_quadrature.fypp

```diff
@@ -14,67 +15,130 @@ module stdlib_experimental_quadrature
     public :: simps
     public :: simps_weights
 
+    ! automatic integration of (weighted) functions
+    #:if QUAD_OK
+    public :: quad
+    public :: weight_t
+    #:for WFUN in WEIGHT_FUNS
+    public :: ${WFUN}$_weight_t
+    #:endfor
+    #:endif
+
 
     interface trapz
-        !! Integrates sampled values using trapezoidal rule
-        !! ([Specification](../page/specs/stdlib_experimental_quadrature.html#description))
-        #:for KIND in REAL_KINDS
-        pure module function trapz_dx_${KIND}$(y, dx) result(integral)
-            real(${KIND}$), dimension(:), intent(in) :: y
-            real(${KIND}$), intent(in) :: dx
-            real(${KIND}$) :: integral
-        end function trapz_dx_${KIND}$
+      !! Integrates sampled values using trapezoidal rule
+      !! ([Specification](../page/specs/stdlib_experimental_quadrature.html#description))
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure module function trapz_dx_${k1}$(y, dx) result(integral)
+          ${t1}$, dimension(:), intent(in) :: y
+          ${t1}$, intent(in) :: dx
+          ${t1}$ :: integral
+        end function trapz_dx_${k1}$      
         #:endfor
-        #:for KIND in REAL_KINDS
-        pure module function trapz_x_${KIND}$(y, x) result(integral)
-            real(${KIND}$), dimension(:), intent(in) :: y
-            real(${KIND}$), dimension(size(y)), intent(in) :: x
-            real(${KIND}$) :: integral
-        end function trapz_x_${KIND}$
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure module function trapz_x_${k1}$(y, x) result(integral)
+            ${t1}$, dimension(:), intent(in) :: y
+            ${t1}$, dimension(size(y)), intent(in) :: x
```

Comment:
Is this a private auxiliary function where you know that the actual argument for `x` will be contiguous in practice, or is this a public user-callable function? (I haven't looked carefully at the whole component to know.) If it is the latter, I don't think you want to be declaring `x` as an explicit-shape array, `dimension(size(y))`. This implies F77-style argument passing, and if the actual argument isn't contiguous it will do an implicit copy-in of that argument to a contiguous temporary and pass it instead. Clearly `x` and `y` need to have the same size, but that style of declaration does nothing whatsoever to ensure that that is the case (in fact it hides the true length of the actual argument). This really requires a run-time check or assertion, with `x` declared as an assumed-shape array.

Reply:
This routine is public by way of the generic `trapz` interface, so your comment is germane. I have no complaints about using assumed-shape declaration here, but I need to educate myself more on array argument passing to understand why it matters. It is a little surprising to learn that the explicit-shape declaration introduces a temporary for non-contiguous actual arguments, whereas the assumed-shape declaration does not.

Reply:
What is described is just the implementation detail of (a) specific compiler(s). Although I didn't have the time to check, I strongly doubt that any of such considerations are present in the standard.

Reply:
It's quite true that this isn't explicitly stated in the standard, however it is implied. The array
```${t1}$, dimension(size(y)), intent(in) :: x``` is simply contiguous (6.5.4) whereas the corresponding actual argument need not be. The only way for that to work is for the array data to be reorganized into contiguous storage.

Reply:
Are you sure that the only possible way to match a non-contiguous array with a contiguous one is by implicit copy-in? I don't know the answer.

Reply:
I think so, but that would be good question for a compiler writer like @pklausler

Reply:
I also wonder if allowing a non-contiguous array into the function (declaring `x` as an assumed-shape array) and working on it, would hurt the overall performance more than a copy-in.

Reply:
I haven't studied the code in question, and may have misunderstood the question; sorry in advance if that's so.

Arguments that have to be contiguous within the called procedure have to be made so by the caller via allocating, populating, passing, copying back, and deallocating a temporary array when the data being passed might not be contiguous.  Arguments that don't have to be contiguous are cheaper to pass using a "descriptor" or "dope vector", but accesses within the called procedure might require more expensive address calculation, and this cost must be paid even when the actual argument is actually contiguous.  But it's not terribly expensive -- this calculation is invariant in any loop and easy to optimize.

The dummy arguments that don't have to be contiguous are non-pointer non-allocatable "assumed-shape" dummy arrays declared with colons for dimensions.

Some actual argument arrays require temporaries to be created for reasons other than contiguity; if these cases are common, the cost of using an assumed-shape dummy argument isn't buying much other than flexibility to handle the others.  For example, passing the result of an array expression.

My general advice (and this is without knowledge of your specific use case, so weight it lightly) is to use assumed-shape dummy arguments for maximum flexibility, with the usual caveat that you should profile execution time later.  If you want to experiment, note that you can still declare an assumed-shape dummy array to be `CONTIGUOUS`, and doing so will help you determine whether some benefit from contiguity actually matters without losing flexibility.  It depends a lot on your target machine, too; SIMD "vectorization" benefits from contiguity, but it may not matter so much on a GPU.

Hope this helps.

Reply:
Thanks for the recommendation! I like the approach of making the in-argument assumed shape, and then later benchmarking against making it contiguous as well. 

A related questions is if function results should also be assumed-shape. E.g.,
````fortran
function trapz_weights_dp(x) result(w)
real(dp), dimension(:), intent(in) :: x
real(dp), dimension(size(x)) :: w ! explicit-shape result
! real(dp), dimension(:), allocatable :: w ! assumed-shape result
...
end function trapz_weights_dp
````
The discussion in #114 suggests that allocatable function results are undesirable. Do I understand right that an explicit-shape result is still expected to be the better option?

Reply:
I'm not sure there was any conclusion in #114 as to which performed better.  The issue probably being whether the array was allocated on the stack or the heap.  I haven't looked to see how you're using this particular function, but personally I've switched to generally avoiding array-valued functions altogether, as I feel this results in a lot of unnecessary allocations/deallocations of memory that passing the result as an argument avoids. I'm not really confident that compiler optimizers are able to eliminate these.

Reply:
@klausler, thanks for that excellent advice and explanation. @milancurcic we should put this recommendation somewhere in our tutorial section or FAQ, as this question will keep coming up often.

Reply:
I have just pushed a commit that makes the `x` argument of `trapz_x_*` and `simps_x_*` assumed-shape with a runtime check on its size. Thanks, @nncarlson, for raising the point, and @epagone and @klausler for the discussion.